File: Codigo2_Python/ListaEncadeada.py
from No import No  # Importa a classe No para utilizar os nós na lista encadeada


class ListaEncadeada:
    def __init__(self):
        # Inicializa a lista encadeada com a cabeça (primeiro nó) como None
        self.cabeca = None

    def adicionar(self, numero, posicao):
        # Método para adicionar um novo nó com o número dado na posição especificada na lista encadeada
        novo = No(numero)  # Cria um novo nó com o número dado
        if self.cabeca is None:
            # Se a lista estiver vazia, define o novo nó como a cabeça da lista
            self.cabeca = novo
        else:
            # Se a lista não estiver vazia, percorre até a posição desejada e insere o novo nó lá
            atual = self.cabeca
            for i in range(posicao-1):
                if atual.get_ponteiro() is None:
                    break  # Se chegar ao final da lista, para
                atual = atual.get_ponteiro()  # Move para o próximo nó
            # Define o próximo do novo nó
            novo.set_ponteiro(atual.get_ponteiro())
            # Define o novo nó como próximo do nó atual
            atual.set_ponteiro(novo)

    def remover(self, numero):
        # Método para remover o nó com o número dado da lista encadeada
        if self.cabeca is None:
            return -1  # Retorna -1 se a lista estiver vazia
        if self.cabeca.get_ponteiro() is None:
            # Se houver apenas um nó na lista e for igual ao número dado, remove-o
            self.cabeca = None
            return -1
        if self.cabeca.get_numero() == numero:
            # Se o primeiro nó contiver o número dado, redefine a cabeça para o próximo nó
            self.cabeca = self.cabeca.get_ponteiro()
            return 1
        atual = self.cabeca
        while atual.get_ponteiro() is not None:
            # Percorre a lista até encontrar o nó com o número dado
            if atual.get_ponteiro().get_numero() == numero:
                atual.set_ponteiro(
                    atual.get_ponteiro().get_ponteiro())  # Remove o nó
                return 1
            atual = atual.get_ponteiro()  # Move para o próximo nó

    def imprimir(self):
        # Método para imprimir os números armazenados na lista encadeada
        lista = []  # Inicializa uma lista vazia para armazenar os números da lista encadeada
        atual = self.cabeca
        while atual is not None:
            # Percorre a lista encadeada e adiciona os números à lista
            lista.append(atual.get_numero())
            atual = atual.get_ponteiro()  # Move para o próximo nó
        print(lista)  # Imprime a lista de números

File: Codigo1_Java/src/ListaEncadeada.java
import java.util.ArrayList;
import java.util.List;

public class ListaEncadeada {
    private No cabeca;

    // Método para adicionar um número em uma posição específica na lista
    public void adicionar(int numero, int posicao) {
        No novo = new No(numero); // Cria um novo nó com o número fornecido
        if (this.cabeca == null) { // Verifica se a lista está vazia
            this.cabeca = novo; // Se estiver vazia, o novo nó é o primeiro da lista
        } else {
            No atual = this.cabeca;
            // Percorre a lista até a posição desejada (posicao - 1)
            for (int i = 0; i < posicao - 1; i++) {
                // Verifica se o próximo nó é nulo para evitar um erro ao tentar acessá-lo
                if (atual.getProximo() == null) {
                    break;
                }
                atual = atual.getProximo(); // Move para o próximo nó
            }
            novo.setProximo(atual.getProximo()); // Define o próximo do novo nó como o próximo do nó atual
            atual.setProximo(novo); // Define o próximo do nó atual como o novo nó
        }
    }

    // Método para remover um número da lista
    public void remover(int numero) {
        if (this.cabeca == null) { // Verifica se a lista está vazia
            return; // Se estiver vazia, não há nada a fazer
        }
        if (this.cabeca.getDado() == numero) { // Verifica se o número está na cabeça da lista
            this.cabeca = this.cabeca.getProximo(); // Se estiver na cabeça, remove a cabeça
            return;
        }
        if (this.cabeca.getProximo() == null) { // Verifica se a lista possui apenas um elemento
            this.cabeca = null; // Se tiver apenas um elemento e não for o número, a lista fica vazia
            return;
        }
        No atual = this.cabeca;
        // Percorre a lista até encontrar o número ou o final da lista
        while (atual.getProximo() != null) {
            if (atual.getProximo().getDado() == numero) { // Verifica se o próximo nó possui o número
                atual.setProximo(atual.getProximo().getProximo()); // Remove o próximo nó
                return;
            }
            atual = atual.getProximo(); // Move para o próximo nó
        }
    }

    // Método para imprimir os elementos da lista
    public void imprimir() {
        List<Integer> lista = new ArrayList<>(); // Cria uma lista auxiliar para armazenar os elementos da lista
                                                 // encadeada
        No atual = this.cabeca;
        // Percorre a lista e adiciona os elementos à lista auxiliar
        while (atual != null) {
            lista.add(atual.getDado());
            atual = atual.getProximo();
        }
        System.out.println(lista); // Imprime a lista auxiliar
    }
}
